File: externals/kyuubi-flink-sql-engine/kyuubi-flink-sql-engine-1.15/src/main/java/org/apache/kyuubi/engine/flink/result/ResultSet.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.kyuubi.engine.flink.result;

import com.google.common.collect.Iterators;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;
import javax.annotation.Nullable;
import org.apache.flink.table.api.ResultKind;
import org.apache.flink.table.api.TableResult;
import org.apache.flink.table.catalog.Column;
import org.apache.flink.table.catalog.ResolvedSchema;
import org.apache.flink.types.Row;
import org.apache.flink.util.Preconditions;
import org.apache.kyuubi.operation.ArrayFetchIterator;
import org.apache.kyuubi.operation.FetchIterator;

/**
 * A set of one statement execution result containing result kind, columns, rows of data and change
 * flags for streaming mode.
 */
public class ResultSet {

  private final ResultKind resultKind;
  private final List<Column> columns;
  private final FetchIterator<Row> data;

  // null in batch mode
  //
  // list of boolean in streaming mode,
  // true if the corresponding row is an append row, false if its a retract row
  private final List<Boolean> changeFlags;

  private ResultSet(
      ResultKind resultKind,
      List<Column> columns,
      FetchIterator<Row> data,
      @Nullable List<Boolean> changeFlags) {
    this.resultKind = Preconditions.checkNotNull(resultKind, "resultKind must not be null");
    this.columns = Preconditions.checkNotNull(columns, "columns must not be null");
    this.data = Preconditions.checkNotNull(data, "data must not be null");
    this.changeFlags = changeFlags;
    if (changeFlags != null) {
      Preconditions.checkArgument(
          Iterators.size((Iterator<?>) data) == changeFlags.size(),
          "the size of data and the size of changeFlags should be equal");
    }
  }

  public List<Column> getColumns() {
    return columns;
  }

  public FetchIterator<Row> getData() {
    return data;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    ResultSet resultSet = (ResultSet) o;
    return resultKind.equals(resultSet.resultKind)
        && columns.equals(resultSet.columns)
        && data.equals(resultSet.data)
        && Objects.equals(changeFlags, resultSet.changeFlags);
  }

  @Override
  public int hashCode() {
    return Objects.hash(resultKind, columns, data, changeFlags);
  }

  @Override
  public String toString() {
    return "ResultSet{"
        + "resultKind="
        + resultKind
        + ", columns="
        + columns
        + ", data="
        + data
        + ", changeFlags="
        + changeFlags
        + '}';
  }

  public static ResultSet fromTableResult(TableResult tableResult) {
    ResolvedSchema schema = tableResult.getResolvedSchema();
    // collect all rows from table result as list
    // this is ok as TableResult contains limited rows
    List<Row> rows = new ArrayList<>();
    tableResult.collect().forEachRemaining(rows::add);
    return builder()
        .resultKind(tableResult.getResultKind())
        .columns(schema.getColumns())
        .data(rows.toArray(new Row[0]))
        .build();
  }

  public static Builder builder() {
    return new Builder();
  }

  /** Builder for {@link ResultSet}. */
  public static class Builder {
    private ResultKind resultKind = null;
    private List<Column> columns = null;
    private FetchIterator<Row> data = null;
    private List<Boolean> changeFlags = null;

    private Builder() {}

    /** Set {@link ResultKind}. */
    public Builder resultKind(ResultKind resultKind) {
      this.resultKind = resultKind;
      return this;
    }

    /** Set columns. */
    public Builder columns(Column... columns) {
      this.columns = Arrays.asList(columns);
      return this;
    }

    /** Set columns. */
    public Builder columns(List<Column> columns) {
      this.columns = columns;
      return this;
    }

    /** Set data. */
    public Builder data(FetchIterator<Row> data) {
      this.data = data;
      return this;
    }

    /** Set data. */
    public Builder data(Row[] data) {
      this.data = new ArrayFetchIterator<>(data);
      return this;
    }

    /** Set change flags. */
    public Builder changeFlags(List<Boolean> changeFlags) {
      this.changeFlags = changeFlags;
      return this;
    }

    /** Returns a {@link ResultSet} instance. */
    public ResultSet build() {
      return new ResultSet(resultKind, columns, data, changeFlags);
    }
  }
}

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/result/ResultSet.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.kyuubi.engine.flink.result

import java.util

import scala.collection.JavaConverters._

import com.google.common.collect.Iterators
import org.apache.flink.api.common.JobID
import org.apache.flink.table.api.{DataTypes, ResultKind}
import org.apache.flink.table.catalog.Column
import org.apache.flink.types.Row

import org.apache.kyuubi.operation.{ArrayFetchIterator, FetchIterator}

case class ResultSet(
    resultKind: ResultKind,
    columns: util.List[Column],
    data: FetchIterator[Row],
    // null in batch mode
    // list of boolean in streaming mode,
    // true if the corresponding row is an append row, false if its a retract row
    changeFlags: Option[util.List[Boolean]]) {

  require(resultKind != null, "resultKind must not be null")
  require(columns != null, "columns must not be null")
  require(data != null, "data must not be null")
  changeFlags.foreach { flags =>
    require(
      Iterators.size(data.asInstanceOf[util.Iterator[_]]) == flags.size,
      "the size of data and the size of changeFlags should be equal")
  }

  def getColumns: util.List[Column] = columns

  def getData: FetchIterator[Row] = data

  def close: Unit = {
    data match {
      case incIte: IncrementalResultFetchIterator => incIte.close()
      case _ =>
    }
  }
}

/**
 * A set of one statement execution result containing result kind, columns, rows of data and change
 * flags for streaming mode.
 */
object ResultSet {

  def fromJobId(jobID: JobID): ResultSet = {
    val data: Array[Row] = if (jobID != null) {
      Array(Row.of(jobID.toString))
    } else {
      // should not happen
      Array(Row.of("(Empty Job ID)"))
    }
    builder
      .resultKind(ResultKind.SUCCESS_WITH_CONTENT)
      .columns(Column.physical("result", DataTypes.STRING()))
      .data(data)
      .build
  }

  def builder: Builder = new ResultSet.Builder

  class Builder {
    private var resultKind: ResultKind = _
    private var columns: util.List[Column] = _
    private var data: FetchIterator[Row] = _
    private var changeFlags: Option[util.List[Boolean]] = None

    def resultKind(resultKind: ResultKind): ResultSet.Builder = {
      this.resultKind = resultKind
      this
    }

    def columns(columns: Column*): ResultSet.Builder = {
      this.columns = columns.asJava
      this
    }

    def columns(columns: util.List[Column]): ResultSet.Builder = {
      this.columns = columns
      this
    }

    def data(data: FetchIterator[Row]): ResultSet.Builder = {
      this.data = data
      this
    }

    def data(data: Array[Row]): ResultSet.Builder = {
      this.data = new ArrayFetchIterator[Row](data)
      this
    }

    def changeFlags(changeFlags: util.List[Boolean]): ResultSet.Builder = {
      this.changeFlags = Some(changeFlags)
      this
    }

    def build: ResultSet = new ResultSet(resultKind, columns, data, changeFlags)
  }
}
